kivy/core/text/_text_pango.pyx

```diff
@@ -0,0 +1,372 @@
+# NOTE: We could probably use a single global FcConfig, but it's not obvious
+# how since adding a font doesn't give you back something that can be directly
+# used to render with. Apparently you must look it up via a font description.
+# Where do you get that from? I don't know... maybe FcConfigGetFonts then
+# pango_fc_font_description_from_pattern(), somehow?
+#
+# The current implementation uses pango_fc_font_map_set_config() to force the
+# loaded TTF to be used, seems to work. As far as I can tell alternative is:
+#
+#     cdef FcConfig *oldconfig = FcConfigGetCurrent()
+#     FcConfigSetCurrent(context.fc_config)
+#     <-- render -->
+#     FcConfigSetCurrent(oldconfig)
+#
+
+cimport cython
+from libc.stdint cimport uint32_t
+from libc.string cimport memset
+from cpython.mem cimport PyMem_Malloc, PyMem_Free
+
+include "../../lib/pangoft2.pxi"
+from kivy.logger import Logger
+from kivy.core.image import ImageData
+
+# Cached contexts; dict key + list value is font filename ("font_name_r")
+# Dict values are ContextContainer instances, one for each font.
+cdef dict kivy_pango_cache = {}
+cdef list kivy_pango_cache_order = []
+
+# Map text direction to pango constant - neutral is considered auto, anything
+# else will call pango_layout_set_auto_dir(the_layout, FALSE)
+cdef dict kivy_pango_text_direction = {
+    'ltr': PANGO_DIRECTION_LTR,
+    'rtl': PANGO_DIRECTION_RTL,
+    'weak_ltr': PANGO_DIRECTION_WEAK_LTR,
+    'weak_rtl': PANGO_DIRECTION_WEAK_RTL,
+    'neutral': PANGO_DIRECTION_NEUTRAL,
+    'auto': PANGO_DIRECTION_NEUTRAL}
+
+
+# Fontconfig and pango structures (one per loaded font). Can't use
+# ctypedef struct here, because that won't fit in the cache dict
+cdef class ContextContainer:
+    cdef PangoContext *context
+    cdef PangoFontMap *fontmap
+    cdef PangoFontDescription *fontdesc
+    cdef PangoFontMetrics *metrics
+    cdef PangoLanguage *metrics_lang # do not free, see docs
+    cdef PangoLayout *layout
+    cdef FcConfig *fc_config
+
+    # Called explicitly on malloc fail to release as fast as possible
```

Comment:
This can be dangerous as you don't have any __cinit__, and so value are not initialized at all. Then __dealloc__ can crash

Reply:
The pointers are guaranteed to be NULL by Cython - cinit not needed here

http://cython.readthedocs.io/en/latest/src/reference/extension_types.html

"By the time __cinit__() is called… All C-level attributes have been initialized to 0 or null."